### PR TITLE
Multicollinearity detection + HDF5 speed up

### LINKED ISSUE #427: Sensitivity Calculation: detection method for multicollinearity for linear regression inside basic statistics post-processor
--------
Issue Description
--------
Migrated from issue 701 on gitlab.

Adding multiple collinear targets when calculating sensitivities results in different results than calculating for each target separately.

See [here](https://hpcgitlab.inl.gov/idaholab/raven/issues/701) for more details.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #931: Slowdown of HDF5 creation after addressing #925
--------
Issue Description
--------
##### What did you expect to see happen?
Ref. #925. The modification to address this issue, determined a slow down of the HDF5 creation.
It needs to be addressed.

##### What did you see instead?
No crash, just a slow down

##### Do you have a suggested fix for the development team?
cache strategy

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)

Closes #427
Closes #931
##### What are the significant changes in functionality due to this change request?
Added a check to detect multicollinearity + accelerate HDF5 creation.
In addition, some utilities functions have been moved to utils (from mathUtils) because some were redundant (e.g. isString in utils and isAString in mathUtils)

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: framework/PostProcessors/BasicStatistics.py [hunk 1]
```diff
@@ -1124,12 +1128,18 @@ def sensitivityCalculation(self,featVars, targVars, featSamples, targSamples, in
...
+            self.raiseAWarning("Condition Number: {:10.4f} > 30.0. Detected SEVERE multicollinearity problem. Sensitivity might be incorrect!".format(condNumber))
           if ind is not None:
             regCoeff = np.insert(regCoeff,ind,1.0)
           senMatrix[p,:] = regCoeff
     else:
       senMatrix = np.zeros((len(targVars), len(featVars)))
       for p, feat in enumerate(featVars):
+        condNumber = np.linalg.cond(featSamples[:,p])
+        if condNumber > 30.:
+          self.raiseAWarning("Condition Number: {:10.4f} > 30.0. Detected SEVERE multicollinearity problem. Sensitivity might be incorrect!".format(condNumber))
```
Comment (wangcj05) [inline framework/PostProcessors/BasicStatistics.py]: There is no need to check it here. The sensitivity is calculated by looping over the feature space, basically, single feature is used here.

Comment (alfoa) [reply framework/PostProcessors/BasicStatistics.py]: removed

**Comment by wangcj05** [comment]
It seems that you pushed the HDF5 changes to this PR. Do you want to combine these changes in one PR? If so, please update the descriptions for this PR and the related issue.

**Comment by wangcj05** [comment]
In addition, the extra check is not removed yet.

**Comment by alfoa** [comment]
The extra check has been removed  (already).

And a new issue is addressed here.

File: framework/h5py_interface_creator.py [hunk 2]
```diff
@@ -302,10 +308,12 @@ def addGroupInit(self,groupName,attributes=None):
...
-    self.h5FileW["allGroupPaths"].resize((len(self.allGroupPaths),))
-    self.h5FileW["allGroupEnds"].resize( (len(self.allGroupEnds),) )
-    self.h5FileW["allGroupPaths"][...] = self.allGroupPaths
-    self.h5FileW["allGroupEnds"][...] = self.allGroupEnds
+    if len(self.allGroupPaths) > len(self.h5FileW["allGroupPaths"]):
+      self.h5FileW["allGroupPaths"].resize((len(self.allGroupPaths)*2,))
+      self.h5FileW["allGroupEnds"].resize( (len(self.allGroupPaths)*2,) )
+    self.h5FileW["allGroupPaths"][len( self.allGroupPaths) - 1] = self.allGroupPaths[-1]
+    self.h5FileW["allGroupEnds"][len(self.allGroupPaths) - 1] = self.allGroupEnds[-1]
+    self.h5FileW.attrs["nGroups"] = len(self.allGroupPaths)
```
Comment (wangcj05) [inline framework/h5py_interface_creator.py]: A function may be useful since these changes are repeated

File: framework/utils/utils.py [hunk 3]
```diff
@@ -390,13 +390,87 @@ def toBytes(s):
...
   """
-    Method to figure out if a variable is a string.
-    @ In, s, object, variable for which we need to assess if it is a string
-    @ Out, isString, bool, true if variable is a str or unicode.
+    Determine if a single-entry value (by traditional standards).
+    Single entries include strings, numbers, NaN, inf, None
+    NOTE that Python usually considers strings as arrays of characters.  Raven doesn't benefit from this definition.
+    @ In, val, object, check
+    @ In, nanOk, bool, optional, if True then NaN and inf are acceptable
+    @ Out, isAScalar, bool, result
```
Comment (wangcj05) [inline framework/utils/utils.py]: ```isAScalar``` -> ```isSingleValued```

File: framework/utils/utils.py [hunk 4]
```diff
@@ -390,13 +390,87 @@ def toBytes(s):
...
+  if isinstance(val,six.integer_types) or  isinstance(val,(float,numpy.number)):
+    # bools are ints, unfortunately
+    if isABoolean(val):
+      return False
+    # nan and inf are floats
+    if nanOk:
+      return True
+    elif val not in [numpy.inf,numpy.nan]:
+      return True
+  return False
```
Comment (wangcj05) [inline framework/utils/utils.py]: you can use functions ```isAFloat``` and ```sAInteger``` instead

File: tests/framework/Databases/HDF5_selective/output_generator.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import math
+import numpy
+
+def run(self, Input):
```
Comment (wangcj05) [inline tests/framework/Databases/HDF5_selective/output_generator.py]: docstring?

File: tests/framework/Databases/HDF5_selective/output_generator.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  number_of_steps = 16
```
Comment (wangcj05) [inline tests/framework/Databases/HDF5_selective/output_generator.py]: remove underscores?

File: tests/framework/Databases/HDF5_selective/output_generator.py (hunk 5 above)
Comment (alfoa) [reply tests/framework/Databases/HDF5_selective/output_generator.py]: done

File: framework/utils/utils.py (hunk 4 above)
Comment (alfoa) [reply framework/utils/utils.py]: done

File: framework/utils/utils.py (hunk 3 above)
Comment (alfoa) [reply framework/utils/utils.py]: done

File: tests/framework/Databases/HDF5_selective/output_generator.py (hunk 6 above)
Comment (alfoa) [reply tests/framework/Databases/HDF5_selective/output_generator.py]: done

**Comment by wangcj05** [comment]
Changes are good to me. Please fix the failed tests. @alfoa

File: framework/h5py_interface_creator.py (hunk 2 above)
Comment (alfoa) [reply framework/h5py_interface_creator.py]: added

**Comment by alfoa** [comment]
@wangcj05 done

**Comment by wangcj05** [comment]
Checklist is good and all tests are green. This PR can be merged.

**Pull request merged**